NuRadioReco/framework/parameter_storage.py

```diff
@@ -0,0 +1,101 @@
+import NuRadioReco.framework.parameters
+
+import itertools
+import logging
+logger = logging.getLogger('NuRadioReco.framework.parameter_storage')
+
+
+class ParameterStorage:
+
+    def __init__(self, parameter_types):
+        self._parameters = {}
+        self._parameter_covariances = {}
+        if not isinstance(parameter_types, list):
+            parameter_types = [parameter_types]
+        self._parameter_types = parameter_types
+
+    def add_parameter_type(self, parameter_type):
+        self._parameter_types.append(parameter_type)
+
+    def __setitem__(self, key, value):
+        self.set_parameter(key, value)
+
+    def __getitem__(self, key):
+        return self.get_parameter(key)
+
+    def _check_key(self, key):
+        if not isinstance(key, tuple(self._parameter_types)):
+            logger.error(f"Parameter {key} needs to be of type {self._parameter_types}")
+            raise ValueError(f"Parameter {key} needs to be of type {self._parameter_types}")
+
+    def get_parameter(self, key):
+        self._check_key(key)
+        return self._parameters[key]
+
+    def has_parameter(self, key):
+        self._check_key(key)
+        return key in self._parameters
+
+    def set_parameter(self, key, value):
+        self._check_key(key)
+        self._parameters[key] = value
+
+    def set_parameter_error(self, key, value):
+        self._check_key(key)
+        self._parameter_covariances[(key, key)] = value ** 2
+
+    def get_parameter_error(self, key):
+        self._check_key(key)
+        return self._parameter_covariances[(key, key)] ** 0.5
+
+    def has_parameter_error(self, key):
+        self._check_key(key)
+        return (key, key) in self._parameter_covariances
+
+    def remove_parameter(self, key):
+        self._check_key(key)
+        self._parameters.pop(key, None)
+
+    def get_parameters(self):
+        return self._parameters
+
+    def serialize(self):
+        parameters = {str(key): self._parameters[key] for key in self._parameters}
+        parameter_covariances = {
+            (str(key[0]), str(key[1])): self._parameter_covariances[key]
+            for key in self._parameter_covariances}
+
+        data = {
+            "_parameters": parameters,
+            "_parameter_covariances": parameter_covariances,
+            "_parameter_types": [parameter_type.__name__ for parameter_type in self._parameter_types]
+        }
+
+        return data
+
+
+    def deserialize(self, data):
+        # for backward compatibility
+        if 'parameters' in data:
+            data['_parameters'] = data['parameters']
+
+        if 'parameter_covariances' in data:
+            data['_parameter_covariances'] = data['parameter_covariances']
```

Comment:
Is there a reason why you are now using `_parameters` to store the parameters, instead of `parameters` as previously? Would just changing it to `parameters` mean you could get rid of this backwards-compatibility check?

Reply:
The first class I touched used `_paramters` I thought all did but realised only later that several (maybe more) used `parameter`. So we need this anyway...